### DISCUSSION TITLE
[python] How would I match inside brackets for an Optional[...] type definition?

### DISCUSSION THREAD

**Discussion by bedecarroll** [ast-grep/ast-grep#1492] [General]
Hi there, big fan of ast-grep. It is exceptionally powerful. Unfortunately I've been having trouble crafting rules that would allow me to rewrite python typing from the old style `List[...]` to the new style `list[...]` and adding `from __future__ import annotations` to the import statement.

I've spent a bit of time reading the documentation but unfortunately I've not been able to figure out how I would accomplish my mission.

My attempts have particularly fallen flat on the changing `Optional[$type_name]` with `$type_name | None`. It seems I can match the `Optional` but as soon as I attempt to capture the type inside brackets I lose the match. This is my (obviously) incorrect attempt.

```
rule:
  pattern: Optional[$A]
  inside:
    kind: type_parameter
    stopBy: end
fix: $A | None
```

I also have successfully matched `List` in a `from typing import List, ...` and with an empty fix I can delete it but it does leave a dangling `,` behind. Is there a way to remove that?

I've been trying out in playground: [playground link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiTGlzdFskQl0iLCJyZXdyaXRlIjoibGlzdFskQl0iLCJzdHJpY3RuZXNzIjoic21hcnQiLCJzZWxlY3RvciI6ImV4cHJlc3Npb25fc3RhdGVtZW50IiwiY29uZmlnIjoiLS0tXG5pZDogYWRkLWFubm90YXRpb25zXG5ydWxlOlxuICBwYXR0ZXJuOiBmcm9tIHR5cGluZyBpbXBvcnQgJCQkQVxuICBraW5kOiBpbXBvcnRfZnJvbV9zdGF0ZW1lbnRcbmZpeDogfC1cbiAgZnJvbSBfX2Z1dHVyZV9fIGltcG9ydCBhbm5vdGF0aW9uc1xuICBmcm9tIHR5cGluZyBpbXBvcnQgJCQkQVxuLS0tXG5pZDogbGlzdFxucnVsZTpcbiAgcGF0dGVybjogTGlzdFxuICBpbnNpZGU6XG4gICAga2luZDogZnVuY3Rpb25fZGVmaW5pdGlvblxuICAgIHN0b3BCeTogZW5kXG5maXg6XG4gIGxpc3Rcbi0tLVxuaWQ6IGRpY3RcbnJ1bGU6XG4gIHBhdHRlcm46IERpY3RcbiAgaW5zaWRlOlxuICAgIGtpbmQ6IGZ1bmN0aW9uX2RlZmluaXRpb25cbiAgICBzdG9wQnk6IGVuZFxuZml4OlxuICBkaWN0XG4tLS1cbmlkOiBvcHRpb25hbFxucnVsZTpcbiAgcGF0dGVybjogT3B0aW9uYWxbJEFdXG4gIGluc2lkZTpcbiAgICBraW5kOiB0eXBlX3BhcmFtZXRlclxuICAgIHN0b3BCeTogZW5kXG5maXg6ICRBIHwgTm9uZVxuLS0tXG5pZDogcmVtb3ZlLW9sZC1pbXBvcnRzXG5ydWxlOlxuICBhbnk6XG4gICAgLSBwYXR0ZXJuOiBMaXN0XG4gICAgLSBwYXR0ZXJuOiBEaWN0XG4gICAgLSBwYXR0ZXJuOiBPcHRpb25hbFxuICBpbnNpZGU6XG4gICAga2luZDogaW1wb3J0X2Zyb21fc3RhdGVtZW50XG4gICAgc3RvcEJ5OiBuZWlnaGJvclxuZml4OiBcIlwiIiwic291cmNlIjoiZnJvbSB0eXBpbmcgaW1wb3J0IExpc3QsIERpY3QsIEFueSwgT3B0aW9uYWxcbmltcG9ydCB0eXBpbmdcbmltcG9ydCB0eXBpbmcgYXMgdFxuXG5kZWYgZm9vKGw6IExpc3Rbc3RyXSwgZDogRGljdFtzdHIsIEFueV0sIG1heWJlOiBPcHRpb25hbFtzdHJdKS0+c3RyOlxuICBwYXNzIn0=)

Any thoughts would be greatly appreciated. Thank you for the great tool!

**Comment by HerringtonDarkholme** [comment]
Hi, thanks for using ast-grep. Let's break the scenarios into several different pieces and break them one by one.

# `Optional[$A]`
> changing Optional[$type_name] with $type_name | None

First ,the `Optional[$TYPE]` will not be interpreted as generic type, but instead a subscript. 

You can see details in the playground tree output and pattern debugger
<img width="1679" alt="image" src="https://github.com/user-attachments/assets/f4f1ee37-2c02-4cdb-bb1e-b2e843e2ddb0">


[Playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicHl0aG9uIiwicXVlcnkiOiJPcHRpb25hbFskQV0iLCJyZXdyaXRlIjoiIiwic3RyaWN0bmVzcyI6InNtYXJ0Iiwic2VsZWN0b3IiOiIiLCJjb25maWciOiItLS1cbmlkOiBhZGQtYW5ub3RhdGlvbnNcbnJ1bGU6XG4gIHBhdHRlcm46IGZyb20gdHlwaW5nIGltcG9ydCAkJCRBXG4gIGtpbmQ6IGltcG9ydF9mcm9tX3N0YXRlbWVudFxuZml4OiB8LVxuICBmcm9tIF9fZnV0dXJlX18gaW1wb3J0IGFubm90YXRpb25zXG4gIGZyb20gdHlwaW5nIGltcG9ydCAkJCRBXG4tLS1cbmlkOiBsaXN0XG5ydWxlOlxuICBwYXR0ZXJuOiBMaXN0XG4gIGluc2lkZTpcbiAgICBraW5kOiBmdW5jdGlvbl9kZWZpbml0aW9uXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDpcbiAgbGlzdFxuLS0tXG5pZDogZGljdFxucnVsZTpcbiAgcGF0dGVybjogRGljdFxuICBpbnNpZGU6XG4gICAga2luZDogZnVuY3Rpb25fZGVmaW5pdGlvblxuICAgIHN0b3BCeTogZW5kXG5maXg6XG4gIGRpY3Rcbi0tLVxuaWQ6IG9wdGlvbmFsXG5ydWxlOlxuICBwYXR0ZXJuOiBPcHRpb25hbFskQV1cbiAgaW5zaWRlOlxuICAgIGtpbmQ6IHR5cGVfcGFyYW1ldGVyXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDogJEEgfCBOb25lXG4tLS1cbmlkOiByZW1vdmUtb2xkLWltcG9ydHNcbnJ1bGU6XG4gIGFueTpcbiAgICAtIHBhdHRlcm46IExpc3RcbiAgICAtIHBhdHRlcm46IERpY3RcbiAgICAtIHBhdHRlcm46IE9wdGlvbmFsXG4gIGluc2lkZTpcbiAgICBraW5kOiBpbXBvcnRfZnJvbV9zdGF0ZW1lbnRcbiAgICBzdG9wQnk6IG5laWdoYm9yXG5maXg6IFwiXCIiLCJzb3VyY2UiOiJcblxuZGVmIGZvbyhtYXliZTogT3B0aW9uYWxbc3RyXSk6IHBhc3MgIn0=)

To fix it, you can use [pattern object](https://ast-grep.github.io/guide/rule-config/atomic-rule.html#pattern-object)

```yml
rule:
  pattern: 
    context: 'a: Optional[$A]'
    selector: generic_type
```

[Updated playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoicHl0aG9uIiwicXVlcnkiOiJhOiBPcHRpb25hbFskQV0iLCJyZXdyaXRlIjoiIiwic3RyaWN0bmVzcyI6InNtYXJ0Iiwic2VsZWN0b3IiOiJnZW5lcmljX3R5cGUiLCJjb25maWciOiItLS1cbmlkOiBhZGQtYW5ub3RhdGlvbnNcbnJ1bGU6XG4gIHBhdHRlcm46IGZyb20gdHlwaW5nIGltcG9ydCAkJCRBXG4gIGtpbmQ6IGltcG9ydF9mcm9tX3N0YXRlbWVudFxuZml4OiB8LVxuICBmcm9tIF9fZnV0dXJlX18gaW1wb3J0IGFubm90YXRpb25zXG4gIGZyb20gdHlwaW5nIGltcG9ydCAkJCRBXG4tLS1cbmlkOiBsaXN0XG5ydWxlOlxuICBwYXR0ZXJuOiBMaXN0XG4gIGluc2lkZTpcbiAgICBraW5kOiBmdW5jdGlvbl9kZWZpbml0aW9uXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDpcbiAgbGlzdFxuLS0tXG5pZDogZGljdFxucnVsZTpcbiAgcGF0dGVybjogRGljdFxuICBpbnNpZGU6XG4gICAga2luZDogZnVuY3Rpb25fZGVmaW5pdGlvblxuICAgIHN0b3BCeTogZW5kXG5maXg6XG4gIGRpY3Rcbi0tLVxuaWQ6IG9wdGlvbmFsXG5ydWxlOlxuICBwYXR0ZXJuOiBPcHRpb25hbFskQV1cbiAgaW5zaWRlOlxuICAgIGtpbmQ6IHR5cGVfcGFyYW1ldGVyXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDogJEEgfCBOb25lXG4tLS1cbmlkOiByZW1vdmUtb2xkLWltcG9ydHNcbnJ1bGU6XG4gIGFueTpcbiAgICAtIHBhdHRlcm46IExpc3RcbiAgICAtIHBhdHRlcm46IERpY3RcbiAgICAtIHBhdHRlcm46IE9wdGlvbmFsXG4gIGluc2lkZTpcbiAgICBraW5kOiBpbXBvcnRfZnJvbV9zdGF0ZW1lbnRcbiAgICBzdG9wQnk6IG5laWdoYm9yXG5maXg6IFwiXCIiLCJzb3VyY2UiOiJcblxuZGVmIGZvbyhtYXliZTogT3B0aW9uYWxbc3RyXSk6IHBhc3MgIn0=)

Relevant doc: 
1. https://ast-grep.github.io/advanced/pattern-parse.html#ambiguous-pattern-code
2. https://ast-grep.github.io/advanced/faq.html#my-pattern-does-not-work-why
3. https://ast-grep.github.io/advanced/pattern-parse.html#use-ast-grep-playground-to-debug-pattern

**Comment by HerringtonDarkholme** [comment]
# import from

this is a quite complicated rewriting. 

```yml
id: remove-old-imports
rule:
  pattern: 'from typing import $$$LIST'
rewriters:
- id: extract
  rule:
    pattern: $A
    not:
      any:
      - regex: ^List$
      - regex: ^Dict$
      - regex: ^Optional$
  fix: $A
transform:
  NEW_LIST:
    rewrite:
      rewriters: [extract]
      source: $$$LIST
      joinBy: ''
fix: from typing import $NEW_LIST
```
[Playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiYTogTGlzdFskQl0iLCJyZXdyaXRlIjoibGlzdFskQl0iLCJzdHJpY3RuZXNzIjoic21hcnQiLCJzZWxlY3RvciI6ImdlbmVyaWNfdHlwZSIsImNvbmZpZyI6ImlkOiByZW1vdmUtb2xkLWltcG9ydHNcbnJ1bGU6XG4gIHBhdHRlcm46ICdmcm9tIHR5cGluZyBpbXBvcnQgJCQkTElTVCdcbnJld3JpdGVyczpcbi0gaWQ6IGV4dHJhY3RcbiAgcnVsZTpcbiAgICBwYXR0ZXJuOiAkQVxuICAgIG5vdDpcbiAgICAgIGFueTpcbiAgICAgIC0gcmVnZXg6IF5MaXN0JFxuICAgICAgLSByZWdleDogXkRpY3QkXG4gICAgICAtIHJlZ2V4OiBeT3B0aW9uYWwkXG4gIGZpeDogJEFcbnRyYW5zZm9ybTpcbiAgTkVXX0xJU1Q6XG4gICAgcmV3cml0ZTpcbiAgICAgIHJld3JpdGVyczogW2V4dHJhY3RdXG4gICAgICBzb3VyY2U6ICQkJExJU1RcbiAgICAgIGpvaW5CeTogJydcbmZpeDogZnJvbSB0eXBpbmcgaW1wb3J0ICRORVdfTElTVCIsInNvdXJjZSI6ImZyb20gdHlwaW5nIGltcG9ydCBMaXN0LCBEaWN0LCBBbnksIE9wdGlvbmFsXG5pbXBvcnQgdHlwaW5nXG5pbXBvcnQgdHlwaW5nIGFzIHRcblxuZGVmIGZvbyhsOiBMaXN0W3N0cl0sIGQ6IERpY3Rbc3RyLCBBbnldLCBtYXliZTogT3B0aW9uYWxbc3RyXSktPnN0cjpcbiAgcGFzcyJ9)

https://ast-grep.github.io/reference/yaml/rewriter.html
 or 
https://ast-grep.github.io/reference/yaml/fix.html#expandend

may help you

**Reply by HerringtonDarkholme** [reply]
Alternative implementation using expandEnd/expandStart

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiYTogTGlzdFskQl0iLCJyZXdyaXRlIjoibGlzdFskQl0iLCJzdHJpY3RuZXNzIjoic21hcnQiLCJzZWxlY3RvciI6ImdlbmVyaWNfdHlwZSIsImNvbmZpZyI6ImlkOiByZW1vdmUtb2xkXG5ydWxlOlxuICBraW5kOiBkb3R0ZWRfbmFtZVxuICByZWdleDogTGlzdHxEaWN0fE9wdGlvbmFsXG4gIGluc2lkZTpcbiAgICBwYXR0ZXJuOiBmcm9tIHR5cGluZyBpbXBvcnQgJEFcbiAgbnRoQ2hpbGQ6XG4gICAgcG9zaXRpb246IDFcbiAgICByZXZlcnNlOiB0cnVlXG5maXg6XG4gIGV4cGFuZFN0YXJ0OlxuICAgIHJlZ2V4OiAnLCdcbiAgdGVtcGxhdGU6ICcnXG4tLS1cbmlkOiByZW1vdmUtb2xkMlxucnVsZTpcbiAga2luZDogZG90dGVkX25hbWVcbiAgcmVnZXg6IExpc3R8RGljdHxPcHRpb25hbFxuICBpbnNpZGU6XG4gICAgcGF0dGVybjogZnJvbSB0eXBpbmcgaW1wb3J0ICRBXG5maXg6XG4gIGV4cGFuZEVuZDpcbiAgICByZWdleDogJywnXG4gIHRlbXBsYXRlOiAnJyIsInNvdXJjZSI6ImZyb20gdHlwaW5nIGltcG9ydCBMaXN0LCBEaWN0LCBBbnksIE9wdGlvbmFsXG5pbXBvcnQgdHlwaW5nXG5pbXBvcnQgdHlwaW5nIGFzIHRcblxuZGVmIGZvbyhsOiBMaXN0W3N0cl0sIGQ6IERpY3Rbc3RyLCBBbnldLCBtYXliZTogT3B0aW9uYWxbc3RyXSktPnN0cjpcbiAgcGFzcyJ9

**Comment by bedecarroll** [comment]
Wow this is absolutely phenomenal stuff @HerringtonDarkholme, thank you! I think this is past what can reasonable expected of an AST parser but there's a case where there are no imports (like so in [playground](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6InB5dGhvbiIsInF1ZXJ5IjoiYTogTGlzdFskQl0iLCJyZXdyaXRlIjoibGlzdFskQl0iLCJzdHJpY3RuZXNzIjoic21hcnQiLCJzZWxlY3RvciI6ImdlbmVyaWNfdHlwZSIsImNvbmZpZyI6ImlkOiByZW1vdmUtb2xkLWltcG9ydHNcbnJ1bGU6XG4gIHBhdHRlcm46ICdmcm9tIHR5cGluZyBpbXBvcnQgJCQkTElTVCdcbnJld3JpdGVyczpcbi0gaWQ6IGV4dHJhY3RcbiAgcnVsZTpcbiAgICBwYXR0ZXJuOiAkQVxuICAgIG5vdDpcbiAgICAgIGFueTpcbiAgICAgIC0gcmVnZXg6IF5MaXN0JFxuICAgICAgLSByZWdleDogXkRpY3QkXG4gICAgICAtIHJlZ2V4OiBeT3B0aW9uYWwkXG4gIGZpeDogJEFcbnRyYW5zZm9ybTpcbiAgTkVXX0xJU1Q6XG4gICAgcmV3cml0ZTpcbiAgICAgIHJld3JpdGVyczogW2V4dHJhY3RdXG4gICAgICBzb3VyY2U6ICQkJExJU1RcbiAgICAgIGpvaW5CeTogJydcbmZpeDogZnJvbSB0eXBpbmcgaW1wb3J0ICRORVdfTElTVCIsInNvdXJjZSI6IiMgYmVmb3JlOlxuIyBmcm9tIHR5cGluZyBpbXBvcnQgTGlzdCwgRGljdCwgT3B0aW9uYWxcbiMgaW1wb3J0IHJlcXVlc3RzXG4jIGFmdGVyOlxuZnJvbSB0eXBpbmcgaW1wb3J0IFxuaW1wb3J0IHJlcXVlc3RzXG5cbmRlZiBmb28obDogTGlzdFtzdHJdLCBkOiBEaWN0W3N0ciwgc3RyXSwgbWF5YmU6IE9wdGlvbmFsW3N0cl0pLT5zdHI6XG4gIHBhc3MifQ==)) but the parser now interprets the import of the next line as part of the preceding statement. For the benefit of anyone else searching I simply ran a `sed -i '' '/from typing import /d'` to delete the import. I tried to figure out a way to always populate `$NEW_LIST` with an `Any` because then the linter could just remove the line if the `Any` was unneeded. It seems like an odd corner case but thought I'd leave it here if it helped someone else.

Once again this is fantastic. Thank you!
